JointJS statemachine and class diagram visualizer

tamsv opened the pull request:
This pull request is intended to implement a single-level statemachine and class diagram visualizer using JointJS.

The visualizer consists of two main parts
  1. Diagram JSON exporter
  2. JavaScript visualizer (using JointJS)
	
The exporter part has unit tests defined covering most of the functionality.

The visualization can be started from the txtUML menu or the context menu:

![txtUML menu](https://i.imgur.com/87Z3yT2.png)

![context menu](https://i.imgur.com/9wb0eoy.png)

The resulting diagrams are then shown in the browser:
![txtUML menu](https://i.imgur.com/vKZ7GFZ.png)

![context menu](https://i.imgur.com/KCB70Vd.png)


There is currently a bug in the internal browser, described in detail in #475. Due to this, currently the default external browser is used for visualization.

Affected by #450
Affected by #475

dobreffandras commented on dev/plugins/hu.elte.txtuml.export.javascript/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,32 @@
+ org.eclipse.core.runtime,
+ hu.elte.txtuml.export.papyrus,
+ org.eclipse.persistence.moxy,
+ org.eclipse.papyrus.uml.diagram.common
```
The dependency includes:
*.elementsarrangers.txtumllayout.LayoutVisualizerManager
*.elementsarrangers.ArrangeException
*.layout.txtuml.TxtUMLLayoutDescriptor
*.wizardz.VisualizeTxtUMLPage
and
*.preferences.PreferencesManager

All of the above mentioned classes can be refactored not to be part of the papyrus component. However we will have to discuss the details.

dobreffandras commented on dev/plugins/hu.elte.txtuml.export.javascript/src/hu/elte/txtuml/export/javascript/Exporter.java:
```diff
@@ -0,0 +1,178 @@
+import hu.elte.txtuml.export.javascript.resources.ResourceHandler;
+import hu.elte.txtuml.export.papyrus.elementsarrangers.ArrangeException;
+import hu.elte.txtuml.export.papyrus.elementsarrangers.txtumllayout.LayoutVisualizerManager;
+import hu.elte.txtuml.export.papyrus.layout.txtuml.TxtUMLLayoutDescriptor;
```
This refactoring can be postponed. An issue will be created later.

tamsv commented on dev/plugins/hu.elte.txtuml.export.javascript/.gitignore:
```diff
@@ -0,0 +1,3 @@
+bin/
+target/
+src/hu/elte/txtuml/export/javascript/resources/res/input.js
```
That was an exported diagram (normaly this file is generated in the target project), it was copied there to help faster development of the visualizer. It was added to .gitignore to prevent an accidental commit. The resource is no longer there but I forgot about the .gitignore file, I will remove it.

tamsv commented on dev/plugins/hu.elte.txtuml.export.javascript/src/hu/elte/txtuml/export/javascript/json/model/cd/ClassNode.java:
```diff
@@ -0,0 +1,147 @@
+	 * @return the type of the class (note: this currently can be "class" or
+	 *         "abstract")
+	 */
+	public String getType() {
```
The type of a digram's node is used in the JavaScript part to determine decorations (eg.: displaying \<\<abstract\>\> ).
I can factor these strings into something like a CDNodeType enum but in the end something has to be seralized to mark this. (a string in the enum)
I chose string representation to help increase the generated file and the JavaScript part readability.

If the above refactoring is not suitable, I can think of the following alternatives:

- collecting different type of nodes into different collections
- representing this as a set of booleans (this isn't really sustainable if many other types will be introduced)

dobreffandras replied:
I like the solution with enums. As you wrote there is an (opportunity)[https://www.tutorialspoint.com/javaexamples/method_enum1.htm] to "save" the strings into the enum. With this solution the readability is kept in the generated JavaScript files and the code quality is higher on the Java side.

djnemeth commented on dev/plugins/hu.elte.txtuml.export.javascript/src/hu/elte/txtuml/export/javascript/resources/res/visualize.html:
```diff
@@ -0,0 +1,50 @@
+	
+	
+	<div>
+		The source and licensing details of the 3rd party libraries which this page uses can be found at their respective websites:
```
I would suggest _used by this page_ instead of _which this page uses_.

djnemeth commented on dev/plugins/hu.elte.txtuml.export.javascript/src/hu/elte/txtuml/export/javascript/resources/res/visualize.html:
```diff
@@ -0,0 +1,50 @@
+</head>
+<body>
+	<!-- selector link holder -->
+	<table>
```
I suggest to add a frame in order to help understand what these lists mean. For example:

<table>
  <tbody>
    <tr align="center">
      <td colspan="2"><b>Visualized diagrams</b></td>
    </tr>
    <tr align="center">
      <td><b>Class</b></td>
      <td><b>State machine</b></td>
    </tr>
    <tr>
      <td>
        <ul>
          <li>com.example.ClassDiagram</li>
        </ul>
      </td>
      <td>
        <ul>
          <li>com.example.SMDiagram1</li>
          <li>com.example.SMDiagram2</li>
          <li>com.example.SMDiagram3</li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

djnemeth commented:
I'll mirror the extra update site and update the _[mirrors.target](https://github.com/ELTE-Soft/txtUML/blob/be03447a37fc1d8579d6c6ecb4127ce7f4a2200f/dev/releng/hu.elte.txtuml.target/mirrors.target)_ soon™.

djnemeth commented:
Jenkins, retest this, please.

dobreffandras commented:
@djnemeth I didn't find a relevant icon of JointJS so I created an icon with a simple JS text see 218af96

tamsv commented on dev/plugins/hu.elte.txtuml.export.javascript/src/hu/elte/txtuml/export/javascript/Exporter.java:
```diff
@@ -0,0 +1,178 @@
+			// informations
+			List<Statement> statements = lvm.getStatementsSet();
+
+			double spacing = 0.8; // default spacing
```
I can't seem to find how to get the default value. Could you elaborate, please?

dobreffandras replied:
It seems you're using the spacing as some kind of scaling for the result boxes. 

The normal behavior of the arranging algorithm is that the space between the individual boxes is exactly the width and height of a box.  The user can adjust this and specify the percentage of these width and height values. The algorithm take this setting into consideration and the resulting box coordinates and dimensions will be set accordingly.

There is an interesting part in the story however. The algorithm generates the layout on a uniform grid, where the boxes are placed on the points of this grid (without any dimension). But in the link arrange phase the width and height of the boxes matter, and the grid has to be refined. To be able to determine the relative dimensions of the boxes each of them is feeded with an initial (desired) pixel value. This introduces a measure, the `pixel-grid-ratio` which defines how many pixels can fit on a grid. This can be queried from the [LayoutVisualizer](https://github.com/ELTE-Soft/txtUML/blob/master/dev/plugins/hu.elte.txtuml.layout.visualizer/src/hu/elte/txtuml/layout/visualizer/algorithms/LayoutVisualize.java#L105)

These whole process results arranged rectangles with normal (width, height) dimensions and the so called pixelDimension (pixelWidth pixelHeight). Their Position is defined on the grid though, so they may be scaled up with the `pixel-grid-ratio`. I'm doing this with the [LayoutTransformer](https://github.com/ELTE-Soft/txtUML/blob/master/dev/plugins/hu.elte.txtuml.export.papyrus/src/hu/elte/txtuml/export/papyrus/elementsarrangers/txtumllayout/LayoutTransformer.java)

dobreffandras replied:
If you have questions or something  is not clear, let me know.

djnemeth replied:
Is this the last remaining task before this pull request can be merged?

dobreffandras replied:
Yes

djnemeth commented:
This is interesting, for some reason Jenkins doesn't build new commits automatically. I'll look into it soon™.

djnemeth commented:
Apparently, the persistence is shared between the JointJS and Papyrus visualization wizards. It's not too big a problem though, perhaps a bit unintuitive.

dobreffandras commented:
Yes, you're right. Currently there are 3 common parts of the two visualization plugins, which has to be separated: Wizards, Persistence, Diagram descriptions. I don't suggest fixing these issues now though, since the current papyrus plugin is outdated, and will be replaced soon. These fixes could cause problems upon merging that branch.

djnemeth commented:
I see. In this case, we could open a new issue regarding these architectural flaws after both pull request are merged.

By the way, what do you mean by shared diagram descriptions and how should we separate them?

dobreffandras commented:
Under diagram description I mean the [TxtUMLLayoutDescriptor](https://github.com/ELTE-Soft/txtUML/blob/master/dev/plugins/hu.elte.txtuml.export.papyrus/src/hu/elte/txtuml/export/papyrus/layout/txtuml/TxtUMLLayoutDescriptor.java) which holds the different diagram descriptions together wihich were selected to be visualized.

dobreffandras commented:
Jenkins, retest this, please.

dobreffandras commented:
Jenkins, retest this, please.

djnemeth commented:
This should not be merged before #480.

tamsv commented:
I've moved the scaling and transforming to the Java side and changed the JavaScript part accordingly.
This resolves the issue due to the misusing of spacing.

dobreffandras commented:
Jenkins, retest this, please.

djnemeth commented:
Jenkins, retest this, please.

djnemeth commented:
The tests are failing because Jenkins is building and testing a temporary merge commit from _master_ (instead of the head of the current branch), to simulate the effects of actually merging the pull request. As a result, failing tests indicate that if merged, this would break _master_. Naturally, we have to merge _master_ into this branch first, and then fix the emerging issues.

dobreffandras commented:
I resolved the build errors. This PR is feature complete and able to merge, but we should wait #480 as @djnemeth  [suggested](https://github.com/ELTE-Soft/txtUML/pull/476#issuecomment-286736637).

dobreffandras commented:
#475 must be retested before merging.

djnemeth commented:
#475 is still [unresolved](https://github.com/ELTE-Soft/txtUML/issues/475#issuecomment-292855661).

dobreffandras commented:
@djnemeth do you approve?

djnemeth commented:
I think the text denoting ports isn't considered while determining the width of classes. For example, see the class diagram generated from the PingPong demo:

![xpingpongclassdiagram](https://cloud.githubusercontent.com/assets/13290204/25315869/be2c22bc-285c-11e7-85a5-c422db558921.png)

Is this easily fixable, or should we finally merge the pull request and resolve this issue later?

devaigergely81 commented:
I suggest to merge and open an issue to be fixed afterwards.

djnemeth commented:
I have opened an issue about the incorrectly displayed ports, see #508.
